pyomo/contrib/gdpopt/algorithm_base_class.py

```diff
@@ -0,0 +1,470 @@
+        """
+        self.config = self.CONFIG(kwds.pop('options', {}),
+                                  preserve_implicit=True)
+        self.config.set_value(kwds)
+
+        # We store bounds, timing info, iteration count, incumbent, and the
+        # expression of the original (possibly nonlinear) objective function.
+        self.LB = float('-inf')
+        self.UB = float('inf')
+        self.unbounded = False
+        self.timing = Bunch()
+        self.initialization_iteration = 0
+        self.iteration = 0
+
+        self.incumbent_boolean_soln = None
+        self.incumbent_continuous_soln = None
+
+        self.original_obj = None
+
+        self.log_formatter = ('{:>9}   {:>15}   {:>11.5f}   {:>11.5f}   '
+                              '{:>8.2%}   {:>7.2f}  {}')
+
+    # Support use as a context manager under current solver API
+    def __enter__(self):
+        return self
+
+    def __exit__(self, t, v, traceback):
+        pass
+
+    def available(self, exception_flag=True):
+        """Solver is always available. Though subsolvers may not be, they will
+        raise an error when the time comes.
+        """
+        return True
+
+    def license_is_valid(self):
+        return True
+
+    def version(self):
+        """Return a 3-tuple describing the solver version."""
+        return __version__
+
+    _metasolver = False
+
+    def solve(self, model, **kwds):
+        """Solve the model.
+
+        Args:
+            model (Block): a Pyomo model or block to be solved
+
+        """
+        # I'm going to be nice for now and intercept with a more informative
+        # error message than what the Config system would give.
+        alg = kwds.pop('algorithm', None)
+        if alg is None:
+            alg = kwds.pop('strategy', None)
+        if alg is not None:
+            raise ValueError("Changing the algorithm in the solve method "
+                             "is not supported for algorithm-specific "
+                             "GDPopt solvers. Either use "
+                             "SolverFactory('gdpopt') or instantiate a "
+                             "solver with the algorithm you want to use.")
+
+        config = self.config(kwds.pop('options', {}),
+                             preserve_implicit=True)
+        config.set_value(kwds)
+
+        with lower_logger_level_to(config.logger, tee=config.tee):
+            self._log_solver_intro_message(config)
+
+            with time_code(self.timing, 'total', is_main_timer=True):
+                results = self._gather_problem_info_and_solve_non_gdps(model,
+                                                                       config)
+                # If it wasn't disjunctive, we solved it
+                if results:
+                    return results
+                else:
+                    # main loop implemented by each algorithm
+                    self._solve_gdp(model, config)
+
+            self._get_final_pyomo_results_object()
```

Comment:
No good reason, I don't think, but because the partially-filled-in results object is already stored on the solver. It gets created early with the `_get_pyomo_results_object_with_problem_info` method. I'm confused about the model references though--there's nothing on that object that references the model, is there?